Run task to clean up cluster, etc.

kholdaway opened the pull request:
No intention to merge this right now, but I want some feedback.

Cleanup algorithm:
Function: `run_inspection_cluster`
- Configures volumes for auto-delete
- Removes owner createVolumePermissions from customer snapshot
- Deletes cloudigrade snapshot

Function: `scale_down_cluster`
- Scales down cluster

**Notes**: You will notice that we have to pass more information along in the celery messages.  Basically, all the information required to setup everything is required to tear it down.  I converted the params to a dictionary to avoid having so many params in the methods. I think its more efficient to just flow the information, rather than go back to AWS and look it up.

infinitewarp commented on cloudigrade/account/tasks.py:
```diff
@@ -44,60 +44,82 @@ def copy_ami_snapshot(arn, ami_id, source_region):
+    volume_information['volume_region'] = region
 
-    enqueue_ready_volume.delay(ami_id, volume_id, region)
+    logger.info(_('{}: volume_id={}, volume_region={}').format(
```
`{}`s #s 🙂

infinitewarp commented on cloudigrade/account/tasks.py:
```diff
@@ -194,14 +217,57 @@ def run_inspection_cluster(messages, cloud='aws'):
+        volume.attach_to_instance(
+            Device=mount_point, InstanceId=ec2_instance_id)
+
+        logger.info(_('{} modify volume {} to auto-delete').format(
```
`{}`s #s 🙂

infinitewarp commented on cloudigrade/account/tasks.py:
```diff
@@ -194,14 +217,57 @@ def run_inspection_cluster(messages, cloud='aws'):
+        aws.remove_snapshot_ownership(customer_snapshot)
+
+        # Delete snapshot_copy
+        logger.info(_('{} delete cloudigrade snapshot copy {}').format(
```
`{}`s #s 🙂

infinitewarp commented on cloudigrade/account/tasks.py:
```diff
@@ -194,14 +217,57 @@ def run_inspection_cluster(messages, cloud='aws'):
+            session,
+            message['customer_snapshot_id'],
+            message['source_region'])
+        aws.remove_snapshot_ownership(customer_snapshot)
```
I _think_ we could move `remove_snapshot_ownership` earlier in the general sequence of tasks to immediately after we start the initial snapshot copy. Once we've started the copy, we may no longer need the permissions, and that means we might not need to pass as much information about the customer (the ARN, the customer's snapshot ID) forward.

kholdaway replied:
At the end of copy_ami_snapshot?

kholdaway replied:
Nope, too soon there.  :)

kholdaway replied:
@infinitewarp We can move it to create_volume function.  The copy snapshot must be complete before we can remove ownership.

ilyavw commented on cloudigrade/account/tasks.py:
```diff
@@ -44,60 +44,83 @@ def copy_ami_snapshot(arn, ami_id, source_region):
     """
     zone = settings.HOUNDIGRADE_AWS_AVAILABILITY_ZONE
-    volume_id = aws.create_volume(snapshot_id, zone)
+    volume_id = aws.create_volume(volume_information['snapshot_copy_id'], zone)
```
Part of `create_volume()` is to verify that the snapshot is ready, next line would be a great place to fire off a new task with a `.delay()` that would remove ownership from from the customers account snapshot since we know that our copy was finished.

ilyavw commented on cloudigrade/account/tasks.py:
```diff
@@ -44,60 +44,83 @@ def copy_ami_snapshot(arn, ami_id, source_region):
+    volume = aws.get_volume(
+        volume_information['volume_id'],
+        volume_information['volume_region'])
     aws.check_volume_state(volume)
```
Next line would be a great time to kick off another new task with a `.delay()` that cleans up our copy of the snapshot, since at this point we know our volume is good.

kholdaway commented:
@infinitewarp @ilyavw Take another look.  I broke cleanup into async tasks (remove ownership from snapshot and delete snapshots).  I fixed tests.

I have not added new tests (yet).  Code coverage is lower that it will be.  Again, looking for high level node before I spend much time on writing the new tests.

infinitewarp commented on cloudigrade/account/tasks.py:
```diff
@@ -161,8 +241,9 @@ def run_inspection_cluster(messages, cloud='aws'):
-        messages (list): A list of dictionary items containing
-            the ami_id and volume_id to be inspected.
+        messages (list): A list of dictionary items containing meta-data (arn,
+            source_region, ami_id, customer_snapshot_id, snapshot_copy_id,
```
I think this comment is a leftover from an earlier stage of refactoring. It looks like `messages` is back down to just `ami_id` and `volume_id`.

infinitewarp commented on cloudigrade/account/tests/test_tasks.py:
```diff
@@ -576,64 +656,78 @@ def test_persist_aws_inspection_cluster_results(self):
-            tasks.HOUNDIGRADE_MESSAGE_READ_LEN
-        )
+    @patch('account.tasks.aws')
+    def test_scale_donw_cluster_success(self, mock_aws):
```
typo 🙂 `donw` ➡️ `down`

infinitewarp commented on cloudigrade/util/tests/test_aws_autoscaling.py:
```diff
@@ -123,3 +123,11 @@ def test_scale_up(self, mock_set_scale):
+
+    @patch('util.aws.autoscaling.set_scale')
+    def test_scale_down(self, mock_set_scale):
+        """Assert scale_up sets the scale to exactly 1."""
```
`s/up/down/` and `s/1/0/`

kholdaway commented:
@infinitewarp Addressed your comments.  Also, handled timing issue where cloudigrade cleans up its snapshot copy before customer snapshot ownership is removed.
